pkg/dns/treecache.go

```diff
@@ -0,0 +1,312 @@
+	cache.m.Lock()
+	defer cache.m.Unlock()
+	node := cache.ensureChildNode(path...)
+	node.ChildNodes[key] = subCache
+}
+
+func (cache *TreeCache) GetEntry(key string, path ...string) (interface{}, bool) {
+	cache.m.RLock()
+	defer cache.m.RUnlock()
+	childNode := cache.GetSubCache(path...)
+	val, ok := childNode.Entries[key]
+	return val, ok
+}
+
+func (cache *TreeCache) GetValuesForPathWithRegex(path ...string) []interface{} {
+	cache.m.RLock()
+	defer cache.m.RUnlock()
+	retval := []interface{}{}
+	nodesToExplore := []*TreeCache{cache}
+	for idx, subpath := range path {
+		nextNodesToExplore := []*TreeCache{}
+		if idx == len(path)-1 {
+			// if path ends on an entry, instead of a child node, add the entry
+			for _, node := range nodesToExplore {
+				if subpath == "*" || subpath == "any" {
+					nextNodesToExplore = append(nextNodesToExplore, node)
+				} else {
+					if val, ok := node.Entries[subpath]; ok {
+						retval = append(retval, val)
+					} else {
+						childNode := node.ChildNodes[subpath]
+						if childNode != nil {
+							nextNodesToExplore = append(nextNodesToExplore, childNode)
+						}
+					}
+				}
+			}
+			nodesToExplore = nextNodesToExplore
+			break
+		}
+
+		if subpath == "*" || subpath == "any" {
+			for _, node := range nodesToExplore {
+				for subkey, subnode := range node.ChildNodes {
+					if !strings.HasPrefix(subkey, "_") {
+						nextNodesToExplore = append(nextNodesToExplore, subnode)
+					}
+				}
+			}
+		} else {
+			for _, node := range nodesToExplore {
+				childNode := node.ChildNodes[subpath]
+				if childNode != nil {
+					nextNodesToExplore = append(nextNodesToExplore, childNode)
+				}
+			}
+		}
+		nodesToExplore = nextNodesToExplore
+	}
+
+	for _, node := range nodesToExplore {
+		for _, val := range node.Entries {
+			retval = append(retval, val)
+		}
+	}
+
+	return retval
+}
+
+func (cache *TreeCache) GetEntries(recursive bool, path ...string) []interface{} {
+	cache.m.RLock()
+	defer cache.m.RUnlock()
+	childNode := cache.GetSubCache(path...)
+	if childNode == nil {
+		return nil
+	}
+
+	retval := [][]interface{}{{}}
+	childNode.appendValues(recursive, retval)
+	return retval[0]
+}
+
+func (cache *TreeCache) DeletePath(path ...string) bool {
+	if len(path) == 0 {
+		return false
+	}
+	cache.m.Lock()
+	defer cache.m.Unlock()
+	if parentNode := cache.GetSubCache(path[:len(path)-1]...); parentNode != nil {
+		if _, ok := parentNode.ChildNodes[path[len(path)-1]]; ok {
+			delete(parentNode.ChildNodes, path[len(path)-1])
+			return true
+		}
+	}
+	return false
+}
+
+func (tn *TreeCache) DeleteEntry(key string, path ...string) bool {
+	tn.m.Lock()
+	defer tn.m.Unlock()
+	childNode := tn.GetSubCache(path...)
+	if childNode == nil {
+		return false
+	}
+	if _, ok := childNode.Entries[key]; ok {
+		delete(childNode.Entries, key)
+		return true
+	}
+	return false
+}
+
+func (tn *TreeCache) appendValues(recursive bool, ref [][]interface{}) {
```

Comment:
Make receiver `cache` consistently everywhere? It is much easier to understand that way.